src/libraries/System.Net.Security/src/System/Net/Security/Pal.OSX/SafeDeleteNwContext.cs

```diff
@@ -0,0 +1,815 @@
+                // not cancellable, we save reference to the pending task. In
+                // case of cancellation, we cancel only waiting on the task, and
+                // on the next ReadAsync call, we don't issue another native
+                // nw_connection_receive call, but rather wait for the pending
+                // task to complete.
+                _pendingAppReceiveBufferFillTask ??= FillAppDataBufferAsync();
+                try
+                {
+                    // Wait for the pending task to complete, which will fill the buffer
+                    await _pendingAppReceiveBufferFillTask.WaitAsync(linkedCts.Token).ConfigureAwait(false);
+                }
+                catch (OperationCanceledException)
+                {
+                    // Throw with correct cancellation token if cancellation was
+                    // requested by user
+                    cancellationToken.ThrowIfCancellationRequested();
+
+                    // otherwise we are tearing down the connection, simulate EOS
+                    Debug.Assert(_shutdownCts.IsCancellationRequested, "Expected shutdown cancellation token to be triggered");
+
+                    ObjectDisposedException.ThrowIf(_disposed, _sslStream);
+                }
+                // other exception types are expected to be fatal and it is okay to not
+                // clear the pending task and rethrow the same exception
+
+                _pendingAppReceiveBufferFillTask = null;
+
+                if (_appReceiveBuffer.ActiveLength == 0)
+                {
+                    if (NetEventSource.Log.IsEnabled()) NetEventSource.Info(this, "ReadAsync returning 0 bytes, end of stream reached");
+                    _appReceiveBufferTcs.TrySetResult(final: true);
+                    return 0; // EOF
+                }
+
+                if (NetEventSource.Log.IsEnabled()) NetEventSource.Info(this, $"ReadAsync filled buffer with {_appReceiveBuffer.ActiveLength} bytes");
+
+                int length = Math.Min(_appReceiveBuffer.ActiveLength, buffer.Length);
+                _appReceiveBuffer.ActiveSpan.Slice(0, length).CopyTo(buffer.Span);
+                _appReceiveBuffer.Discard(length);
+                if (NetEventSource.Log.IsEnabled()) NetEventSource.Info(this, $"Read {length} bytes");
+                return length;
+            }
+        }
+
+        internal Task FillAppDataBufferAsync()
+        {
+            bool success = _appReceiveBufferTcs.TryGetValueTask(out ValueTask valueTask, this, CancellationToken.None);
+            Debug.Assert(success, "Concurrent FillAppDataBufferAsync detected");
+
+            if (NetEventSource.Log.IsEnabled()) NetEventSource.Info(this, $"Waiting for read from connection");
+            unsafe
+            {
+                Interop.NetworkFramework.Tls.NwConnectionReceive(ConnectionHandle, StateHandle, 16 * 1024, StateHandle, &CompletionCallback);
```

Comment:
Seems like we're passing StateHandle twice in this call, and that's the only call to this function. We can delete one of the params.